### PR TITLE
Handle sparse matrices with RcppEigen

### PR DESCRIPTION
**Pull request by mnwright**

### CODE REVIEW

File: R/predict.R [hunk 1]
```diff
@@ -185,7 +186,12 @@ predict.ranger.forest <- function(object, data, predict.all = FALSE,
   }
 
   ## Convert to data matrix
-  data.final <- data.matrix(data.used)
+  if (is.matrix(data.used) | inherits(data.used, "Matrix")) {
```
Comment (dselivanov) [inline R/predict.R]: better to have atomic "or": `||`

File: R/ranger.R [hunk 2]
```diff
@@ -334,7 +346,7 @@ ranger <- function(formula = NULL, data = NULL, num.trees = 500, mtry = NULL,
                               data.selected[-1]))
     colnames(data.final) <- c(dependent.variable.name, status.variable.name,
                               independent.variable.names)
-  } else if (is.matrix(data.selected)) {
+  } else if (is.matrix(data.selected) | inherits(data.selected, "Matrix")) {
```
Comment (dselivanov) [inline R/ranger.R]: same `||`

**Comment by dselivanov** [comment]
Hi, thank you very much for this feature. 
> It seems to be a little slower compared to using dense matrices. The reason is probably that element-wise access is not optimal in sparse matrices.

I briefly checked modified files, but since I'm not familiar with ranger internals, I can't figure out patter of access to elements of sparse matrix. Is it row-wise or column-wise? (or by sample or by feature). I suspect it is row-wise, but I can be wrong. In case of row-wise patter, mb it worth to use CSR matrix (`Matrix::dgRMatrix`)? Also I think it is possible to use `Eigen::MappedSparseMatrix` instead of `Eigen::MappedSparseMatrix` - see details [here](https://github.com/RcppCore/RcppEigen/blob/master/inst/include/RcppEigenWrap.h).  But haven't used it as well. As I understand this allows to avoid copies and have only single sparse matrix in memory.
Hope this can help.

**Comment by mnwright** [comment]
Thanks for your comments! Element access is column-wise. However, down in the trees we need only small parts of the columns and therefore single element access is used. It might be possible to optimize this in the case of sparse data. 

`Matrix::dgRMatrix` is not working with `RcppEigen`, so I couldn't test that. I've tested `Eigen::MappedSparseMatrix` and couldn't see any difference in speed.

**Comment by mnwright** [comment]
Is the `RcppEigen` dependency known to cause any problems?

**Pull request merged**